Mage/src/main/java/mage/util/CardUtil.java

```diff
@@ -1696,23 +1705,31 @@ private static Map<String, Object> getCostsTags(Game game, Ability source) {
      * @return if the tag was found
      */
     public static boolean checkSourceCostsTagExists(Game game, Ability source, String tag) {
-        Map<String, Object> costTags = getCostsTags(game, source);
+        Map<String, Object> costTags = getSourceCostsTagsMap(game, source);
         return costTags != null && costTags.containsKey(tag);
     }
     /**
      * Find a specific tag in the cost tags of either the source ability, or the permanent source of the ability.
      * Works in any moment (even before source ability activated)
+     * Do not use with null values, use checkSourceCostsTagExists instead
      *
      * @param game
      * @param source
      * @param tag The tag's string identifier to look up
      * @param defaultValue A default value to return if the tag is not found
      * @return The object stored by the tag if found, the default if not
      */
-    public static Object getSourceCostsTag(Game game, Ability source, String tag, Object defaultValue){
-        Map<String, Object> costTags = getCostsTags(game, source);
+    public static <T> T getSourceCostsTag(Game game, Ability source, String tag, T defaultValue){
+        Map<String, Object> costTags = getSourceCostsTagsMap(game, source);
         if (costTags != null) {
-            return costTags.getOrDefault(tag, defaultValue);
+            Object value = costTags.getOrDefault(tag, defaultValue);
+            if (value == null) {
+                throw new IllegalStateException("Wrong code usage: Costs tag " + tag + " has value stored of type null but is trying to be read. Use checkSourceCostsTagExists");
+            }
+            if (value.getClass() != defaultValue.getClass()) {
```

Comment:
does it not work to check `value instanceof T`, or do you explicitly want to exclude subclasses?

Reply:
`instanceof` needs a concrete class and doesn't work with generics. I could use `getClass().isAssignableFrom(class)` but it feels like the generics might start getting very messy in that situation- the defaultValue would need to be the superclass to begin with which makes using that rather awkward.

Reply:
All good as is then